torch/_functorch/aot_autograd.py

```diff
@@ -918,10 +918,27 @@ def aot_module_simplified(
             aot_config,
         )
 
+    if isinstance(mod, torch._dynamo.utils.GmWrapper):
+        # When we have list inputs to the graph, this function is called by the flatten_graph_inputs
+        # wrapper, which boxes the inputs so that they can be freed before the end of this scope
+        def boxed_forward(runtime_args: List[torch.Tensor]):
+            flat_args = []
+            flat_args.extend(params_flat)
+            flat_args.extend(runtime_args)
+            runtime_args.clear()
+            return compiled_fn(flat_args)
+
+        # Just for convenience
+        boxed_forward.zero_grad = mod.zero_grad
+        boxed_forward.named_parameters = mod.named_parameters
+        boxed_forward.named_buffers = mod.named_buffers
+        return boxed_forward
```

Comment:
I am a bit worried about us regressing at some point: there have been a lot of changes to AOTAutograd in the last year that have involved adding a new wrapper in the runtime logic, and if we ever forget to propagate the list / clear it, we'll go back to not freeing activations early.

It would be great to have a test that would fail if we fail to do this list clearing properly. Absent any other ideas, one option I can think of is a test like this:
```
def assertion_impl(x):
    mem_allocated = torch.cuda.memory_allocated()
    self.assertTrue(mem_allocated < 1000000, "large tensor should have been freed")
    return x.clone()
    
lib = torch.library.Library('test_ns', "FRAGMENT")
lib.define(f"assertion_op(Tensor x) -> Tensor", tags=(torch.Tag.pt2_compliant_tag,))
lib.impl(name, assertion_impl, "CUDA")
lib.impl(name, lambda x: x.clone(), "Meta")

@torch.compile
def f(x):
    new_shape = min(x.shape[0], 2)
    out = torch.ones(new_shape)
    return assertion_op(out)

# allocate a giant cuda tensor, that should be freed half-way through the invocation of the compiled fn    
out = f(torch.ones(1000000, device='cuda'))
```

Reply:
Yes, I agree with the need of tests for this clearing behavior. 

I was thinking of adding checks for the refcounts of stolen inputs, but it would require us adding test hooks and other logic to determine expected refcount. Your example is much less intrusive, I wasn't aware we could define test ops like that!

Reply:
>  I wasn't aware we could define test ops like that!

Yeah... written as above, it's pretty sketchy 😛. Probably what we should do is define that custom op as an assert, and use the effect-token logic to functionalize it when we trace it. But in this case, I think it's probably ok for the test since we don't have to worry about it getting reordered (it needs to run after we finish using `x`).